proposals/0133-syscall-get-epoch-stake.md

```diff
@@ -0,0 +1,96 @@
+---
+simd: '0133'
+title: Syscall Get-Epoch-Stake
+authors:
+  - Joe Caulfield (Anza)
+category: Standard
+type: Core
+status: Draft
+created: 2024-03-25
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+A syscall to retrieve a vote account's delegated stake for the current epoch.
+
+## Motivation
+
+Currently, on-chain programs have no knowledge of the current epoch's stake and
+how much active stake is delegated to any vote account.
+
+If this data were available for querying by on-chain programs, it would unblock
+many use cases, such as validator governance and secondary consensus mechanisms,
+that were previously not possible on Solana.
+
+Additionally, this would enable the Feature Gate program defined in
+[SIMD 0089](./0089-programify-feature-gate.md) to tally vote account stake in
+support for a pending feature gate.
+
+## Alternatives Considered
+
+[SIMD 0056](https://github.com/solana-foundation/solana-improvement-documents/pull/56)
+proposes using an on-chain sysvar account to store all of the current epoch's
+stake.
+
+Because account data is finite, using a sysvar account to store even just the 
+current epoch's stake limits the number of entries that can be stored. The
+amount of validators in one Solana cluster could surpass this number in the
+future.
+
+Exposing epoch-stake information through a syscall avoids this account maximum
+size constraint. While the syscall approach does not offer the easy off-chain
+retrieval of a sysvar, there are existing ways to get epoch-stake data off
+chain. The priority of a new design should be making the data available to
+on-chain programs.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+The specification for the proposed syscall is as follows:
+
+```c
+/**
+ * Retrieves the total active stake delegated to a vote account for the current
+ * epoch.
+ *
+ * @param var_addr      VM memory address to copy the retrieved data to.
+ * @param vote_address  A pointer to 32 bytes representing the vote address.
+ * @return              A 64-bit unsigned integer error code:
+ *                        - 0 if the operation is successful.
+ *                        - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_epoch_stake(
+  /* r1 */ uint8_t *    var_addr,
+  /* r2 */ void const * vote_address,
+);
+```
+
+`var_addr` must be the starting address of 8 bytes of writable VM memory to
+store the `u64` response. If not, the syscall will abort the VM with an access
+violation.
+
+If the provided vote address corresponds to an account that is not a vote
+account or does not exist, the syscall will write `0` for active stake. 
+
```

Comment:
```suggestion
account or does not exist, the syscall will write `0` for active stake. 

This syscall consumes `syscall_base_cost` compute units always.

```

Reply:
Seems like it should maybe be `syscall_base_cost + cost_of(u64)`.

Currently, we charge `syscall_base_cost + cost_of(T)` to write sysvar `T` to VM memory.. Should this `u64` be the same?

Reply:
Gotcha. I was taking a look at sysvars, which is `sysvar_base` plus one CU per byte written.
https://github.com/solana-labs/solana/blob/e2d34d375bad87a95cdfbe4b07163bb13515bfeb/programs/bpf_loader/src/syscalls/sysvar.rs#L10-L16

Note `sysvar_base` and `syscall_base` are the same (`100`) right now, for what it's worth.

I was also checking out some that use length, like `set_return_data`.
https://github.com/solana-labs/solana/blob/e2d34d375bad87a95cdfbe4b07163bb13515bfeb/programs/bpf_loader/src/syscalls/mod.rs#L1249-L1255

I think the latter makes the most sense for 127. Here we can use a similar concept but make things static since it's always going to read a 32-byte address and return an 8-byte integer. Maybe:

```
syscall_base + (32 / cpi_per_u) + (8 / cpi_per_u)
```

Reply:
Just wanted to point out that as per SIMD 0127, this calculation _could_ be:

```
max(sysvar_base, (32 / cpi_per_u) + (length / cpi_per_u))
```

But in this case length is always `8`, so we'd get `max(100, 0.16)`, so we can just charge `(32 / cpi_per_u) + (length / cpi_per_u)` constantly.

Reply:
With the above being said, as well as https://github.com/solana-foundation/solana-improvement-documents/pull/127#discussion_r1546374161:

```
syscall_base + (32/cpi_per_u) + mem_op_base
```